Add dnd_character exercise

mr-sigma opened the pull request:
Add `dnd_character` exercise based on 48in24 implementation status page: https://exercism.org/challenges/48in24/implementation_status

And on the forums: https://forum.exercism.org/t/add-dnd-character-exercise/9492

kotp commented on exercises/practice/dnd-character/.meta/example.rb:
```diff
@@ -0,0 +1,23 @@
+    @wisdom = roll_dice
+    @charisma = roll_dice
+
+    @hitpoints = 10 + self.class.modifier(constitution)
```
This is known from the constitution and we can derive the adjustment from base 10 hitpoints.

We might want to test through the story of the character and not the story of "modifier" as a function in a language.

kotp commented:
I converted to draft as it is still in progress.

kotp commented on exercises/practice/dnd-character/.meta/example.rb:
```diff
@@ -0,0 +1,53 @@
+module Modifiable
+
+  # This only calculates hit point modifier based on constitution
+  def modifier(constitution)
```
Let's ask for the "generic" thing.

mr-sigma replied:
Using `ability_score` as the new argument based of 5e naming conventions: https://roll20.net/compendium/dnd5e/Ability%20Scores#content

kotp commented on exercises/practice/dnd-character/.meta/example.rb:
```diff
@@ -0,0 +1,53 @@
+    @wisdom       = Die.roll
+    @charisma     = Die.roll
+
+    @hitpoints    = BASE_HITPOINTS + self.class.modifier(constitution)
```
And give, as it is here, the more specific thing.

mr-sigma replied:
Using `ability_score` as the new argument based of 5e naming conventions: https://roll20.net/compendium/dnd5e/Ability%20Scores#content

mr-sigma replied:
Wasn't sure what you were asking for here, the modifier method takes an ability score, and in this case we want to calculate it from constitution, so I don't believe this needs any naming changes based on the use case.

kotp replied:
I may have been looking at some other version, you are correct.

kotp commented on exercises/practice/dnd-character/.meta/example.rb:
```diff
@@ -0,0 +1,53 @@
+
+end
+
+if $PROGRAM_NAME == __FILE__
+  puts DndCharacter.allocate.public_methods(false)
+end
```
We can remove the library guard once we are done exploring.  Or at least make it something that is more "example" based rather than "debugging" like.

mr-sigma replied:
Removed!

kotp commented on exercises/practice/dnd-character/dnd_character_test.rb:
```diff
@@ -0,0 +1,52 @@
+
+module Helpers
+  BASE_HITPOINTS = 10
+  MODIFIER_TO_CONSTITUTION = {
```
Constitution modifier to affect hitpoints, so this could be something that may be a different name.

mr-sigma replied:
Gave a more specific name: `CONSTITUTION_MODIFIER_TO_CONSTITUTION_MAP`

kotp replied:
Not a great name.  But naming is hard.  And the `_MAP` is a "type indicator" in the name, which is not common in Ruby.

Let me look at this and see if I can come up with a better name.

kotp commented on exercises/practice/dnd-character/dnd_character_test.rb:
```diff
@@ -0,0 +1,52 @@
+    character = DndCharacter.new
+    constitution_modifier = character.hitpoints - BASE_HITPOINTS
+
+    assert_includes MODIFIER_TO_CONSTITUTION[constitution_modifier], character.constitution
```
We can add a message to the user if it fails that indicates what the result should have been.

kotp commented on exercises/practice/dnd-character/dnd_character_test.rb:
```diff
@@ -0,0 +1,52 @@
+  def test_attributes_being_accessed_do_not_change
+    character = DndCharacter.new
+
+    (attributes << :hitpoints).each do |attribute|
+      assert_equal character.send(attribute), character.send(attribute)
+    end
```
How many times should we test this to have some assurance that they are stable/unchanging?

mr-sigma replied:
As above, thank you for addressing this in the RFC!

kotp commented on exercises/practice/dnd-character/dnd_character.rb:
```diff
@@ -0,0 +1 @@
+
```
We are giving an empty file.  There is commonly a comment here that describes (for those that may have been away for a while) to write the tests in this file, and to read the Instructions or README if they are developing locally.

mr-sigma replied:
Added boilerplate language for the stub file!

kotp commented on exercises/practice/dnd-character/dnd_character_test.rb:
```diff
@@ -0,0 +1,56 @@
+  def test_attributes_being_accessed_do_not_change
+    character = DndCharacter.new
+
+    (attributes << :hitpoints).each do |attribute|
+      assert_equal character.send(attribute), character.send(attribute)
+    end
```
I do not have much confidence that this is doing well what the test says it is doing.

```ruby
    (attributes << :hitpoints).each do |attribute|
      expected = character.send(attribute)
      10.times do
        assert_equal expected, character.send(attribute)
      end
```
This would check each 10 (arbitrary choice) times, against what it was the answer for the first call

kotp commented on exercises/practice/dnd-character/dnd_character_test.rb:
```diff
@@ -0,0 +1,56 @@
+
+module Helpers
+  BASE_HITPOINTS = 10
+  CONSTITUTION_MODIFIER_TO_CONSTITUTION_MAP = {
```
This table seems to be `ABILITY_SCORE_TO_MODIFIER`.

kotp commented on exercises/practice/dnd-character/dnd_character.rb:
```diff
@@ -0,0 +1,6 @@
+=begin
+Write your code for the 'D&D Character' exercise in this file. Make the tests in
+`dnd_character_test.rb` pass.
```
```suggestion

`dnd_character_test.rb` pass.
```

kotp commented on exercises/practice/dnd-character/dnd_character.rb:
```diff
@@ -0,0 +1,6 @@
+=begin
+Write your code for the 'D&D Character' exercise in this file. Make the tests in
+`dnd_character_test.rb` pass.
+To get started with TDD, see the `README.md` file in your
```
```suggestion
`dnd_character_test.rb` pass.

To get started with TDD, see the `README.md` file in your
```

iHiD commented:
@mr-sigma @kotp Unless I'm missing something, the tests.toml and the tests themselves don't seem to be aligned. Compare this to C# for example: https://github.com/exercism/csharp/blob/main/exercises/practice/dnd-character/DndCharacterTests.cs

kotp commented:
> @mr-sigma @kotp Unless I'm missing something, the tests.toml and the tests themselves don't seem to be aligned. Compare this to C# for example: https://github.com/exercism/csharp/blob/main/exercises/practice/dnd-character/DndCharacterTests.cs

The tests are not one to one in alignment, this is true.  And CSharp having an exact one to one based on Problem Specification makes sense to me, since it is being generated, and it appears without any additional changes or removals from the Problem Specifications supplied information.

We should find that the tests do not need to be exact matches one to one, and may even be difficult to match, given the random nature of character generation, against a user written design, without likely driving a specific design or solution (and this is a practice exercise currently.)

The tests focus on the modification algorithm being applied, and the tests that attempt to ensure that the character attributes are changing randomly per character being generated, but also unchanging for the character once it is created.

We should find that the tests align in result, though, since deviation from that would be problematic if and when we decide to automate the tests, for example, and when bringing in problem descriptions from upstream.

Also, I think this pull request should still be in "Draft Mode" so we are still working through some things regarding the test and how it influences the solution.

iHiD commented:
My understanding is that tests should align 1-1 with the tests.toml. 

I've looked through many other tracks in case I misunderstood this, and they are all 1-1 to problem specs:

- https://github.com/exercism/javascript/blob/main/exercises/practice/dnd-character/dnd-character.spec.js
- https://github.com/exercism/bash/blob/main/exercises/practice/dnd-character/dnd_character.bats
- https://github.com/exercism/elixir/blob/main/exercises/practice/dnd-character/test/dnd_character_test.exs
- https://github.com/exercism/wren/blob/main/exercises/practice/dnd-character/dnd-character.spec.wren

I'll wait for @ErikSchierboom to weigh in on this on Tuesday, but my understanding of exercises Exercism-wide is that they should 1-1 align to tests.toml.

kotp commented:
> My understanding is that tests should align 1-1 with the tests.toml.
> 
> I've looked through many other tracks in case I misunderstood this, and they are all 1-1 to problem specs:
> 
> * https://github.com/exercism/javascript/blob/main/exercises/practice/dnd-character/dnd-character.spec.js
> * https://github.com/exercism/bash/blob/main/exercises/practice/dnd-character/dnd_character.bats
> * https://github.com/exercism/elixir/blob/main/exercises/practice/dnd-character/test/dnd_character_test.exs
> * https://github.com/exercism/wren/blob/main/exercises/practice/dnd-character/dnd-character.spec.wren
> 
> I'll wait for @ErikSchierboom to weigh in on this on Tuesday, but my understanding of exercises Exercism-wide is that they should 1-1 align to tests.toml.

I will defer.  When I was creating the generator, there were exercises that did not necessarily match up with Ruby, and I had made exceptions, but that was years ago, so it could have changed.  I understand that many are like that, I just was not aware of the requirement to do so other than it being convenient to do so.  (Some of the generators would conditional do things bringing in some tests, and making others or not do others.)

kotp commented on exercises/practice/dnd-character/dnd_character_test.rb:
```diff
@@ -0,0 +1,102 @@
+        assert_includes allowed_range, character.send(attribute)
+      end
+
+      assert_equal expected_hitpoints, character.hitpoints
```
This one is in the correct order, the expected first, and then character hitpoints being actual.

Indeed, we could remove the `expected_` from the name, or we could say only `expected` because the actual shows what kind of thing is expected based on the method being called on the object under test.

mr-sigma commented:
@kotp I added your patches, rebased, and ran `rubocop -d -A` to align with the repo's style guidelines. Should be all set to go, thanks for the feedback!

kotp reviewed (approved):
I sent a patch, but am approving early.

I will accept your choice on the what you see there and what you choose to apply.

This is definitely looking good.

mr-sigma commented:
@kotp, patch applied!

kotp commented on exercises/practice/dnd-character/dnd_character_test.rb:
```diff
@@ -0,0 +1,119 @@
+
+      attributes.each do |attribute|
+        assert_includes allowed_range, character.send(attribute),
+          informative_message % [attribute, allowed_range] # rubocop:disable Style/FormatString
```
```suggestion
          informative_message % [attribute, allowed_range]
```

kotp commented on exercises/practice/dnd-character/dnd_character_test.rb:
```diff
@@ -0,0 +1,119 @@
+      informative_message = %q(The character's %s must be %s)
+
+      assert_equal expected_hitpoints, character.hitpoints,
+        informative_message % ['hitpoints', expected_hitpoints] # rubocop:disable Style/FormatString
```
```suggestion
        informative_message % ['hitpoints', expected_hitpoints]
```

kotp reviewed (changes requested):
Remove redundant inline disables.

kotp commented:
Thank you very much @mr-sigma
